File: concepts/big-o/loglinear-time-example/javascript/mergeSort.js
function mergeSort(arr) {
  // base case
  if (arr.length < 2) {
    return arr;
  }

  // divide array into two
  const leftArr = arr.slice(0, arr.length / 2);
  const rightArr = arr.slice(arr.length / 2);

  // sort each half array
  const sortedLeft = mergeSort(leftArr);
  const sortedRight = mergeSort(rightArr);

  // merge sorted left and right arrays
  const sortedArr = mergeSortedArrays(sortedLeft, sortedRight);

  return sortedArr;
}

function mergeSortedArrays(arr1, arr2) {
  let i = 0;
  let j = 0;
  const sortedArr = [];

  while (i < arr1.length && j < arr2.length) {
    if (arr1[i] < arr2[j]) {
      sortedArr.push(arr1[i]);
      i += 1;
    } else {
      sortedArr.push(arr2[j]);
      j += 1;
    }
  }

  while (i < arr1.length) {
    sortedArr.push(arr1[i]);
    i += 1;
  }

  while (j < arr2.length) {
    sortedArr.push(arr2[j]);
    j += 1;
  }

  return sortedArr;
}

const arr = [2, 4, 9, 6, 1, 8, 3, 5, 7];
// [ 1, 2, 3, 4, 5, 6, 7, 8, 9]
console.log(mergeSort(arr));

File: concepts/big-o/loglinear-time-example/java/MergeSort.java
import java.util.List;
import java.util.ArrayList;

public class MergeSort {
    public static List<Integer> mergeSort(List<Integer> list) {
        // base case - already sorted
        if (list.size() < 2) {
            return list;
        }

        // otherwise divide list into two halves
        List<Integer> leftList = new ArrayList<>(list.subList(0, list.size() / 2));
        List<Integer> rightList = new ArrayList<>(list.subList(list.size() / 2, list.size()));

        // sort each half
        List<Integer> sortedLeft = mergeSort(leftList);
        List<Integer> sortedRight = mergeSort(rightList);

        // merge sorted halves
        return merge(sortedLeft, sortedRight);
    }

    private static List<Integer> merge(List<Integer> left, List<Integer> right) {
        List<Integer> merged = new ArrayList<>();

        int leftIndex = 0;
        int rightIndex = 0;

        // merge the two arrays while they each have elements
        while (leftIndex < left.size() && rightIndex < right.size()) {
            if (left.get(leftIndex) < right.get(rightIndex)) {
                merged.add(left.get(leftIndex));
                leftIndex += 1;
            } else {
                merged.add(right.get(rightIndex));
                rightIndex += 1;
            }
        }

        // append any remaining elements
        while (leftIndex < left.size()){
            merged.add(left.get(leftIndex));
            leftIndex += 1;
        }

        while (rightIndex < right.size()){
            merged.add(right.get(rightIndex));
            rightIndex += 1;
        }

        return merged;
    }

    public static void main(String[] args) {
        List<Integer> arr = List.of(2, 4, 9, 6, 1, 8, 3, 5, 7);
        List<Integer> sortedArr = mergeSort(arr);
        // [1, 2, 3, 4, 5, 6, 7, 8, 9]
        System.out.println(sortedArr);
    }
}